README.md
# clip-test

This library is a subset of `clojure.test` for Clojure and ClojureScript. It
lets you write tests (albeit with constraints) that can run on both Clojure and
ClojureScript.

You can also use this library to write JVM-specific and CLJS-specific tests.


## Usage

For ClojureScript testing this library should be used in conjunction with
[lein-cljsbuild](https://github.com/emezeske/lein-cljsbuild).

On Clojars: On Clojars: https://clojars.org/clip-test

Leiningen dependency: `[clip-test "0.2.2"]`

Supported Clojure versions: 1.2, 1.3, 1.4, 1.5

Tested with lein-cljsbuild version: 0.2.9

### Notes:

* Only `deftest`, `testing` and `is` macros are supported.
* `thrown?` and `thrown-with-msg?` are implemented as macros; unlike
  `clojure.test`, there is no multi-method to extend the dispatch on.
* You must implement the `test-ns-hook` function in the test namespace and call
  it externally to run the tests in CLJS.
* Write the tests in the same style as in `test/clip_test/core_test.clj`.
* Invoke the tests in CLJS in a way similar as in `run-tests.js`/`run-tests.html`
  and `test-cljs/clip_test/run_tests.cljs`.
* You need to have PhantomJS (`run-tests.js`) and/or
  Firefox+Firebug (`run-tests.html`) or any browser supporting JavaScript console
  to run the tests in CLJS.
* Printing of JavaScript stack trace on CLJS test errors is supported only on
  PhantomJS 1.6+ and modern browsers.
* The namespaces `clip-test.testutil` and `clip-test.testutil-cljs` have common
  functions that provide similar functionality in Clojure and ClojureScript.

### Examples

The following projects use _clip-test_ for CLJS testing:

* [Quiddity](https://github.com/kumarshantanu/quiddity)
* [Basil](https://github.com/kumarshantanu/basil)

## Development of this library

### Running tests

Clean old stuff if required (esp for ClojureScript testing):

```
lein clean
lein dev cljsbuild clean
```

Run the tests (you may see failures and errors - they are expected):

```
lein dev test           # test with latest stable Clojure version
lein dev cljsbuild test # test with ClojureScript (needs `phantomjs` installed)
lein all test           # test with all supported Clojure versions
```

### Building the JAR

Clean old stuff if required (see 'Running tests' section above.) Then run this:

```
lein jar      # create the JAR file
lein install  # install JAR to local Maven repo
```

## Getting in touch

On Twitter: [@kumarshantanu](https://twitter.com/kumarshantanu)

E-mail: kumar(dot)shantanu at gmail.com

## License

Copyright © 2012 Shantanu Kumar

Distributed under the Eclipse Public License, the same as Clojure.


CHANGES.md
# CHANGES and TODO

## 2012-October-21 / 0.2.2

* Print JS stack trace on CLJS test errors (PhantomJS 1.6+ or modern browser)

## 2012-October-21 / 0.2.1

* Print namespace when running tests

## 2012-October-16 / 0.2.0

* Utility functions: `millis-now` `sleep`

## 2012-August-30 / 0.1.0

* Macros: `deftest` `testing` `is`
* Special macros: `thrown?` `thrown-with-msg?`
* Explicit reporting of test result summary
* Utility functions: `re-quote` `throw` `try-catch` `char?` `read-str`

run-tests.js
try {
  phantom.injectJs('target/clip-test.js');
  console.log('Injected target/clip-test.js');
  clip_test.run_tests.run();
} catch (e) {
  console.log('Found exception');
  console.log(e);
  console.log(e.fileName);
  console.log(e.trace);
} finally {
  phantom.exit();
}

test/clip_test/jvm_test.clj
(ns clip-test.jvm-test
  (:use [clip-test.testutil
         :only [read-str re-quote throw-msg try-catch error-msg]])
  (:use [clip-test.core
         :only [deftest testing is]]))


(deftest test-success
  (testing "single, success test"
    (is true "success")))


(deftest test-failure
  (testing "single, failure test"
    (is false "failure")))


(deftest test-error
  (testing "error test"
    (is (thrown? RuntimeException
           (throw-msg "foo")) "failure")
    (is (thrown-with-msg? RuntimeException #"foo"
           (throw-msg "foo")) "failure with message")))


(defn test-ns-hook
  []
  (test-success)
  (test-failure)
  (test-error))

test/clip_test/core_test.clj
(ns clip-test.core-test
  (:use [clip-test.testutil;*CLJSBUILD-REMOVE*;-cljs
         :only [;*CLJSBUILD-REMOVE*;RuntimeException
                re-quote throw-msg millis-now sleep]])
  (:use;*CLJSBUILD-REMOVE*;-macros
    [clip-test.core;*CLJSBUILD-REMOVE*;-cljs
     :only [deftest testing is
            ;*CLJSBUILD-REMOVE*;thrown? thrown-with-msg?
            ]]))


(deftest test-success
  (is true)
  (is true "without 'testing' wrapper")
  (testing "is truthy"
    (is true)
    (is true "true"))
  (testing "is ="
    (is (= 10 10))
    (is (= 10 10) "= 10 10"))
  (testing "is thrown?"
    (is (thrown? Throwable
                 (assert (not "Sample message"))))
    (is (thrown? Throwable
                 (assert (not "Sample message"))) "thrown? Throwable"))
  (testing "is thrown-with-msg? (re-quote)"
    (is (thrown-with-msg? RuntimeException (re-quote "Sample message")
                          (throw-msg "Sample message")))
    (is (thrown-with-msg? RuntimeException (re-quote "Sample message")
                          (throw-msg "Sample message"))
        "thrown-with-msg? (re-quote)"))
  (testing "is thrown-with-msg? (regex)"
    (is (thrown-with-msg? RuntimeException #"Sample.*"
                          (throw-msg "Sample message")))
    (is (thrown-with-msg? RuntimeException #"Sample.*"
                          (throw-msg "Sample message"))
        "thrown-with-msg? (regex)")))


(deftest test-fail
  (is false)
  (is false "without 'testing' wrapper")
  (testing "is falsy"
    (is false)
    (is false "false"))
  (testing "is inequal"
    (is (= 2 3))
    (is (= 2 3) "= 2 3")))


(deftest test-error
  (testing "is error"
    (is (throw-msg "Raising error"))
    (is (throw-msg "Raising error") "throw-msg")))


(deftest test-testutil
  (testing "millis-now"
    (is (pos? (millis-now)) "millis-now"))
  (testing "sleep"
    (let [start (millis-now)
        _     (sleep 100)
        stop  (millis-now)]
    (is (>= (- stop start) 100) "sleep"))))


(defn test-ns-hook
  []
  (test-success)
  (test-fail)
  (test-error)
  (test-testutil))

src/clip_test/internal.clj
(ns clip-test.internal
  "Internal stuff for use by the namespace clip-test.core-cljs. Some of these
  must be re-initialized in a platform-specific way before they can be used.")


(defn try-catch
  "Handler fn that takes 2 args `try-f` (argless fn) and `catch-f` (fn that
  takes the exception object as only arg).
  NOTE: This function should be re-initialized in a platform-specific way."
  [try-f catch-f]
  (assert (not "set! `try-catch` as per documentation")))


(defn error-msg
  "Return the error message from the given platform-specific error object.
  NOTE: This function should be re-initialized in a platform-specific way."
  [err]
  (assert (not "set! `error-message` as per documentation")))


;; ----- BEGIN: CLJS specific stuff -----


(def ^:dynamic *test-name* nil)
(def ^:dynamic *testing*   nil)
(def ^:dynamic *assertion* nil)

(def test-count   (atom 0))
(def assert-count (atom 0))
(def fail-count   (atom 0))
(def error-count  (atom 0))


(defn ++
  [atom]
  (swap! atom inc))


(defn print-msg
  [msg status expected actual]
  (print (str "        " status (when msg (str " in {" msg "}"))))
  (print (str "        expected: " (pr-str expected) "\n"
              "          actual: " (pr-str actual))))


(defn fail
  [msg expected actual]
  (++ fail-count)
  (print-msg msg "FAIL" expected actual))


(defn error
  [msg expected err]
  (++ error-count)
  (print-msg msg "ERROR" expected err))


;; ----- END: CLJS specific stuff -----

src/clip_test/testutil.clj
(ns clip-test.testutil
  (:require [clip-test.internal :as internal])
  (:import [java.util.regex Pattern]))


(def read-str read-string)


(defn re-quote
  [x] {:pre [(string? x)]}
  (re-pattern (Pattern/quote x)))


(defn throw-msg
  [x] {:pre [(string? x)]}
  (throw (RuntimeException. ^String x)))


(defn try-catch
  [try-f catch-f]
  (try (try-f)
    (catch Throwable t
      (catch-f t))))


(defn error-msg
  [err] {:pre [(instance? Throwable err)]}
  (.getMessage ^Throwable err))


(defn millis-now
  []
  (System/currentTimeMillis))


(defn sleep
  [millis] {:pre [(number? millis)]}
  (try (Thread/sleep millis)
    (catch InterruptedException e
      (.interrupt ^Thread (Thread/currentThread)))))


(defn init!
  []
  (alter-var-root #'internal/try-catch (constantly try-catch))
  (alter-var-root #'internal/error-msg (constantly error-msg)))


(init!)

src/clip_test/core_cljs.clj
;*CLJSBUILD-MACRO-FILE*;

(ns clip-test.core-cljs
  (:use [clip-test.internal :only [try-catch error-msg
                                   *test-name* *testing* *assertion*
                                   test-count assert-count ++ fail error]]))


(defmacro deftest
  [the-name & body]
  `(defn ~the-name
     []
     (binding [*test-name* ~(str the-name)]
       (++ test-count)
       (print (str "[Running test] " ~(str (ns-name *ns*)) "/" ~(str the-name)))
       ~@body)))


(defmacro testing
  [msg & body]
  `(binding [*testing* ~msg]
     (print (str "  -> Testing '" ~msg "'"))
     ~@body
     (print (str "  `~~~~~~~~~"))))


(defmacro thrown?
  "set! `try-catch` before calling this, so that the effect is same as:
  `(try ~@body
     (catch ~'js/Error err#
       true))"
  [ex-class & body]
  `(try-catch #(do ~@body false)
              (constantly true)))


(defmacro thrown-with-msg?
  "set! `try-catch` before calling this, so that the effect is same as:
  `(try ~@body
     (catch ~'js/Error err#
       ...))"
  [ex-class msg-regex & body]
  `(try-catch #(do ~@body false)
              (fn [err#]
                (let [r# (if (string? ~msg-regex)
                           (= ~msg-regex (error-msg err#))
                           (boolean (re-find ~msg-regex (error-msg err#))))]
                  (when-not r#
                    (fail *assertion* ~msg-regex err#))
                  true))))


(defmacro is
  ([form] `(is ~form nil))
  ([form msg]
    `(do (++ assert-count)
       (binding [*assertion* ~msg]
         (try-catch #(let [value# ~form]
                       (when-not value#
                         (fail ~msg '~form value#)))
                    (fn [err#]
                      (error ~msg '~form err#)
                      (print (or (.-stack err#)
                                 (.-stacktrace err#)
                                 (.-stackTrace err#)))))))))


project.clj
(defproject clip-test "0.2.2"
  :description "Subset of `clojure.test` for Clojure and ClojureScript"
  :url "https://github.com/kumarshantanu/clip-test"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  ;; :dependencies [[org.clojure/clojure "1.4.0"]]
  :warn-on-reflection true
  :plugins [[lein-cljsbuild "0.2.9"]]
  :profiles {:jst {:source-paths ["src" "test"]
                   ;; Enable the lein hooks for: clean, compile, test, and jar
                   ;; :hooks [leiningen.cljsbuild]
                   :cljsbuild {:crossovers [clip-test.core-test
                                            clip-test.internal]
                               ;; Command for running the unit tests in CLJS
                               ;;     $ lein cljsbuild test
                               :test-commands {"unit" ["phantomjs"
                                                       "run-tests.js"]}
                               :builds {:test {:source-path "test-cljs"
                                               :compiler
                                               {:output-to "target/clip-test.js"
                                                ;; :optimizations nil
                                                :pretty-print true}}}}}
             :1.2 {:dependencies [[org.clojure/clojure "1.2.1"]]}
             :1.3 {:dependencies [[org.clojure/clojure "1.3.0"]]}
             :1.4 {:dependencies [[org.clojure/clojure "1.4.0"]]}
             :1.5 {:dependencies [[org.clojure/clojure "1.5.0-alpha7"]]}}
  :aliases {"all" ["with-profile" "1.2:1.3:1.4:1.5"]
            "dev" ["with-profile" "1.4,jst"]})
